coala_quickstart/generation/SettingsClass.py

```diff
@@ -0,0 +1,256 @@
+        for the classes to store metadata.
+        :param key:
+            The setting value as a string.
+        :param func:
+            The function object. Either create_arguments() for linter bears
+            or run() for other bears.
+        :param bear:
+            The bear object.
+        """
+        present_in_annot = in_annot(func, key)
+        present_in_default_args = in_default_args(func, key)
+        present_in_all_args = in_all_args(func, key)
+
+        if present_in_annot:
+            self.diff_bool_others(key, present_in_annot)
+        elif present_in_all_args and not present_in_default_args:
+            self.diff_bool_others(key, get_all_args(func)[key])
+        else:
+            self.parse_dep_tree(bear, key)
+
+    def parse_dep_tree(self, bear, key):
+        """
+        Parses the bear's dependency tree looking for
+        non-optional setting and their Type.
+        :param bear:
+            The bear object.
+        :param key:
+            The setting value as a string.
+        """
+        deps = bear.BEAR_DEPS
+        for dep in deps:
+            present_in_annot = in_annot(dep.run, key)
+            if present_in_annot:
+                self.diff_bool_others(key, present_in_annot)
+            else:
+                settings = get_all_args(dep.run)
+                for pointer in get_default_args(dep.run):
+                    del settings[pointer]
+                if key in settings:
+                    self.diff_bool_others(key, settings[key])
+            self.parse_dep_tree(dep, key)
+
+    def diff_bool_others(self, key, check):
+        """
+        Checks if a settings is of Type bool or any other based
+        on the value of check and feeds
+        up the classes to store metadata.
+        :param key:
+            The Setting value as a string.
+        :param check:
+            The Type to which the key is to be checked against.
+        """
+
+        if check == bool:
+            self.settings_bool.append(key)
+        else:
+            self.settings_others.append(key)
+
+    def diff_bool_others_default(self, key, check):
+        """
+        Checks if a settings is of Type bool or any other based
+        on the value of check and feeds
+        up the metaclass.
+        :param key:
+            The Setting value as a string.
+        :param check:
+            The Type to which the key is to be checked against.
+        """
+        if isinstance(check, bool):
+            self.settings_bool.append(key)
+        else:
+            self.settings_others.append(key)
+
+
+class BearSettings:
+
+    """
+    Collect optional and non-optional settings for each bear
+    """
+
+    def __init__(self, bear):
+        """
+        :param bear:
+            A bear class object.
+        """
+        self.bear = bear
+        function = bear.create_arguments if 'linter' in str(bear) else bear.run
+        function_name = 'create_arguments' if 'linter' in str(bear) else 'run'
```

Comment:
Ideally, we should be using the function name of the `function` variable assigned in the line above

Reply:
It was difficult to extract the name in case of decorated functions, although didn't try using `__name__` property